extensions-core/kinesis-indexing-service/src/main/java/org/apache/druid/indexing/kinesis/KinesisRecordSupplier.java

```diff
@@ -0,0 +1,693 @@
+                  recordBufferFullWait
+              );
+
+              shardIterator = kinesis.getShardIterator(
+                  record.getStream(),
+                  record.getPartitionId(),
+                  ShardIteratorType.AT_SEQUENCE_NUMBER.toString(),
+                  record.getSequenceNumber()
+              ).getShardIterator();
+
+              rescheduleRunnable(recordBufferFullWait);
+              return;
+            }
+          }
+
+          shardIterator = recordsResult.getNextShardIterator(); // will be null if the shard has been closed
+
+          rescheduleRunnable(fetchDelayMillis);
+        }
+        catch (ProvisionedThroughputExceededException e) {
+          long retryMs = Math.max(PROVISIONED_THROUGHPUT_EXCEEDED_BACKOFF_MS, fetchDelayMillis);
+          rescheduleRunnable(retryMs);
+        }
+        catch (Throwable e) {
+          log.error(e, "getRecordRunnable exception, retrying in [%,dms]", EXCEPTION_RETRY_DELAY_MS);
+          rescheduleRunnable(EXCEPTION_RETRY_DELAY_MS);
+        }
+
+      };
+    }
+
+    private void rescheduleRunnable(long delayMillis)
+    {
+      if (started && !stopRequested) {
+        try {
+          scheduledExec.schedule(getRecordRunnable(), delayMillis, TimeUnit.MILLISECONDS);
+        }
+        catch (RejectedExecutionException e) {
+          log.info(
+              "Rejecting fetch records runnable submission, worker for partition[%s] is not in a valid state",
+              streamPartition.getPartitionId()
+          );
+
+        }
+      } else {
+        log.info("Worker for partition[%s] has been stopped", streamPartition.getPartitionId());
+      }
+    }
+  }
+
+  // used for deaggregate
+  private final MethodHandle deaggregateHandle;
+  private final MethodHandle getDataHandle;
+
+  private final AmazonKinesis kinesis;
+
+  private final int recordsPerFetch;
+  private final int fetchDelayMillis;
+  private final boolean deaggregate;
+  private final int recordBufferOfferTimeout;
+  private final int recordBufferFullWait;
+  private final int fetchSequenceNumberTimeout;
+  private final int maxRecordsPerPoll;
+  private final int fetchThreads;
+  private final int recordBufferSize;
+
+  private ScheduledExecutorService scheduledExec;
+
+  private final Map<StreamPartition<String>, PartitionResource> partitionResources = new ConcurrentHashMap<>();
+  private BlockingQueue<OrderedPartitionableRecord<String, String>> records;
+
+  private volatile boolean checkPartitionsStarted = false;
+  private volatile boolean closed = false;
+
+  public KinesisRecordSupplier(
+      AmazonKinesis amazonKinesis,
+      int recordsPerFetch,
+      int fetchDelayMillis,
+      int fetchThreads,
+      boolean deaggregate,
+      int recordBufferSize,
+      int recordBufferOfferTimeout,
+      int recordBufferFullWait,
+      int fetchSequenceNumberTimeout,
+      int maxRecordsPerPoll
+  ) throws ClassNotFoundException, IllegalAccessException, NoSuchMethodException
```

Comment:
This doesn't need to throw these checked exceptions. Please return `RuntimeException` instead.